### PR TITLE
SS17 Final Code Changes

### PR DESCRIPTION
**Pull request by byteofprash**
1. All the code part of the Final presentation is pushed. 
2. Doxygen comments for the modules added
3. Documentation updated.
4. ROS messages updated, with description. 
5. RosMsgUtil template with Services and msgs implemented. 

**setup.sh** script will install all the dependencies for

### CODE REVIEW

**Comment by missxa** [comment]
o_O

**Comment by byteofprash** [comment]
Yes? Is something missing?

**Comment by missxa** [comment]
lol, what's happening?
no, it's just huge
don't know if there's smth missing yet

File: .idea/misc.xml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (missxa) [inline .idea/misc.xml]: remove .idea folder

File: src/vision_service/srv/DescribeScene.srv [hunk 2]
```diff
@@ -0,0 +1,3 @@
+---
```
Comment (missxa) [inline src/vision_service/srv/DescribeScene.srv:1]: src/vision_service is not needed. remove the folder and mention in the docs, that roboy_communication_cognition is required

File: src/zed/zedRoboy.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+from ctypes import *
```
Comment (missxa) [inline src/zed/zedRoboy.py:1]: what is it? we are planning to use the ZED official python wrapper. seems like zed-related cpp/hpp files are obsolete. delete unnecessary or explain the need

File: src/ObjectRecognition.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
+"""@package RecogniseFace
```
Comment (missxa) [inline src/ObjectRecognition.py]: typo

File: src/RecogniseFace.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
+"""@package RecogniseFace
```
Comment (missxa) [inline src/RecogniseFace.py:1]: typo

File: src/RecogniseFace.py [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+    phase_train_placeholder = graph.get_tensor_by_name("phase_train:0")
+    embeddings = graph.get_tensor_by_name("embeddings:0")
+    print('done.')
+    print("Time taken:",time.time()-start," Seconds")
+    vs = cv2.VideoCapture(0)
+    facesEmbeddings = []
+    if not vs:
+        sys.exit()
+    while True:
+        prashFace = [-0.00111124,  0.05805413,  0.02128297,  0.03982484,  0.12203307,
```
Comment (missxa) [inline src/RecogniseFace.py]: store features in a separate file and load them (pickle/json)

File: src/RecogniseFace.py [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+            facesEmbeddings = []
+            # for i in range(len(FaceVal)):
+            # print("i is:",i)
+            aligned_face, lm = align_face_dlib(image, FaceVal[0], AlignDlib.OUTER_EYES_AND_NOSE)
+
+            feed_dict = {
+                image_batch: np.expand_dims(aligned_face, 0),
+                phase_train_placeholder: False
+            }
+           # train(session)
```
Comment (missxa) [inline src/RecogniseFace.py]: clean up comments

File: src/RosMsgUtil.py [hunk 8]
```diff
@@ -1,9 +1,142 @@
...
+1. This module has Util functions to send out ROS messages.
+2. The send functions here are called from different modules.
+"""
+import websocket
+import json as json
+import numpy as np
+ws = websocket.WebSocket();
+ws.connect('ws://localhost:9090')   
+
+def AdvertiseNewFacialFeatures():
```
Comment (missxa) [inline src/RosMsgUtil.py:11]: duplicated functionality with NewFacialFeatures.msg

File: src/RosMsgUtil.py [hunk 9]
```diff
@@ -1,9 +1,142 @@
...
+                          \"topic\": \"/roboy/cognition/vision/FaceCoordinates\"\
+                        }")   
+
+def AdvertiseFindObject():
+    ws.send("{ \"op\": \"advertise_service\",\
+                          \"type\": \"roboy_communication_cognition/FindObjects\",\
+                          \"service\": \"/roboy/cognition/vision/FindObjects\"\
+                        }")  
+
+def AdvertiseDescribeScene():
```
Comment (missxa) [inline src/RosMsgUtil.py:29]: duplicated functionality with DescribeSceneSrv.py

File: src/RosMsgUtil.py [hunk 10]
```diff
@@ -1,9 +1,142 @@
...
-	ws = websocket.WebSocket();
-	ws.connect("ws://bot.roboy.org:9090")
-	ws.send("""data:'{"receiver":"/roboy/cognition/vision/faces","msg":"{data:''}","type":"package/Type"}'""")
-	print ("Sent")
-	print ("Reeiving...")
-	ws.close()	
\ No newline at end of file
+"""@package RosMsgUtil
+1. This module has Util functions to send out ROS messages.
+2. The send functions here are called from different modules.
```
Comment (missxa) [inline src/RosMsgUtil.py:3]: could not find RosMsgUtil function calls in other modules

File: src/RosMsgUtil.py [hunk 11]
```diff
@@ -1,9 +1,142 @@
...
+                        }")  
+
+def AdvertiseLookAtSpeaker():
+    ws.send("{ \"op\": \"advertise_service\",\
+                          \"type\": \"roboy_communication_cognition/LookAtSpeaker\",\
+                          \"service\": \"/roboy/cognition/vision/LookAtSpeaker\"\
+                        }")  
+
+
+def AdvertiseContinously():
```
Comment (missxa) [inline src/RosMsgUtil.py:42]: does not need to be continuous, just once
+typo in name

**Review by missxa — CHANGES_REQUESTED** [review]
Since Github webview lags due to insane amount of files changed in this PR (probably): here's docs review summary 
1. Update index.rst, add what are current capabilities of the Vision module
2.  Update getting started to include how to run it and use your module
3.  Move architecture diagram from solution strategy into a separate chapter
4. Add a link from "getting started" to public interfaces. rename it to Public Interfaces (ROS)
5. Add API section: has to show your in-code comments (i don't understand why you didn't do it yet, since the comments are there). here is what i mean http://roboy-memory.readthedocs.io/en/develop/ScopeContext/9_API.html

Files in repo:
- remove .idea folder
- or any __pycache__ folders you pushed to the repo
- remove .catkin_workspace
- update your .gitignore to avoid this kind of things in the future

As I can see, ROS functionality is there but is not currently used. We'll work on it starting next semester then. Please address the comments so we do a release asap

**Comment by byteofprash** [comment]
@missxa Could we proceed with the merging into master? Is something else pending apart from the API docs? If API docs is the only thing pending, let's have that as a separate PR and merge these.

**Review by missxa — APPROVED** [review]

**Pull request merged**